nameko/containers.py

```diff
@@ -353,41 +348,55 @@ def spawn_worker(self, entrypoint, args, kwargs,
-        gt = self._worker_pool.spawn(self._run_worker, worker_ctx,
-                                     handle_result)
-        self._active_threads[gt] = entrypoint
-        gt.link(self._handle_thread_exited)
+        gt = self._worker_pool.spawn(
+            self._run_worker, worker_ctx, handle_result
+        )
+        self._worker_threads[worker_ctx] = gt
+        self.worker_created(worker_ctx)
+        gt.link(self._handle_worker_thread_exited, worker_ctx)
         return worker_ctx
 
-    def spawn_managed_thread(self, run_method, protected=False):
-        """ Spawn a managed thread to run ``run_method``.
-
-        Threads can be marked as ``protected``, which means the container will
-        not forcibly kill them until after all extensions have been killed.
-        Extensions that require a managed thread to complete their kill
-        procedure should ensure to mark them as ``protected``. For example,
-        :class:`nameko.messaging.QueueConsumer` cleanly closes connections
-        to the AMQP broker even when killed; the thread that holds that
-        connection is ``protected`` so it isn't stopped until after the
-        QueueConsumer returns from its kill.
+    def spawn_managed_thread(self, fn, extension=None, protected=None):
+        """ Spawn a managed thread to run ``fn`` on behalf of `extension`.
 
-        Any uncaught errors inside ``run_method`` cause the container to be
-        killed.
+        Any uncaught errors inside ``fn`` cause the container to be killed.
 
         It is the caller's responsibility to terminate their spawned threads.
         Threads are killed automatically if they are still running after
         all extensions are stopped during :meth:`ServiceContainer.stop`.
 
         Extensions should delegate all thread spawning to the container.
         """
-        gt = eventlet.spawn(run_method)
-        if not protected:
-            self._active_threads[gt] = MANAGED_THREAD
-        else:
-            self._protected_threads.add(gt)
-        gt.link(self._handle_thread_exited)
+        if not isinstance(extension, Extension):
+            warnings.warn(
+                "The signature of `spawn_managed_thread` has changed. "
+                "The `protected` kwarg is now deprecated, and extensions "
+                "should pass themselves as the second argument for better "
+                "logging. See :meth:`nameko.containers.ServiceContainer."
+                "spawn_managed_thread`.", DeprecationWarning
+            )
+        if extension is None:
+            extension = "<unknown-extension>"
+
+        gt = eventlet.spawn(fn)
+        self._managed_threads[gt] = extension
+        gt.link(self._handle_managed_thread_exited, extension)
         return gt
 
+    def worker_created(self, worker_ctx):
+        """ Called after a worker is created.
+
+        This is a hook for `ServiceContainer` subclasses to track worker
+        creation.
```

Comment:
Struggling to come up with something that's helpful here. It's a hook for you to do what you want with... I don't think there's anything specific you really _shouldn't_ do